tapdb/addrs.go

```diff
@@ -681,22 +684,41 @@ func (t *TapAddressBook) InsertScriptKey(ctx context.Context,
 	scriptKey asset.ScriptKey, declaredKnown bool) error {
 
 	var writeTxOpts AddrBookTxOptions
-	return t.db.ExecTx(ctx, &writeTxOpts, func(q AddrBook) error {
+	err := t.db.ExecTx(ctx, &writeTxOpts, func(q AddrBook) error {
 		internalKeyID, err := insertInternalKey(
 			ctx, q, scriptKey.RawKey,
 		)
 		if err != nil {
 			return fmt.Errorf("error inserting internal key: %w",
 				err)
 		}
+
 		_, err = q.UpsertScriptKey(ctx, NewScriptKey{
 			InternalKeyID:    internalKeyID,
 			TweakedScriptKey: scriptKey.PubKey.SerializeCompressed(),
 			Tweak:            scriptKey.Tweak,
 			DeclaredKnown:    sqlBool(declaredKnown),
 		})
-		return err
+		if err != nil {
+			return fmt.Errorf("error inserting script key: %w", err)
+		}
+
+		// This key might already exist, so we'll make sure that
+		// declared_known is set to true. Otherwise, if the key already
+		// existed, and we need to view it as ours, we'll fail to do so.
+		if declaredKnown {
+			return q.DeclareScriptKeyKnown(
```

Comment:
Argh, another oversight on my part... An alternative fix (and perhaps the one expected) could be to update the `ON CONFLICT` clause to set

```
  -- This is not a NOP, we overwrite the declared_known value.
  DO UPDATE SET  tweaked_script_key = EXCLUDED.tweaked_script_key,
                               declared_known = EXCLUDED.declared_known
```

Reply:
I was thinking the same - in that case idk if a separate query would be useful?

Reply:
Mhmm, we could also go a step further and also allow a tweak to be updated as well. 

IIUC, we have some other logic that skips an asset in the default balance set if the asset has a script key tweak. In my testing, that wasn't triggering as we'd insert the transfer for the force close (insert script keys w/o tweak), then I'd want to insert the script keys again before trying to make contract resolutions (which failed to update). 

My rationale for doing it with this new method was that it only affected the call path I was concerned with. Will switch to the more general method and see if that breaks anything elsewhere. Agree the upsert extension is cleaner.